### ISSUE TITLE
multiarch support for packages

### ISSUE THREAD

**Issue by rn** [linuxkit/linuxkit#2292]
We now have both x86_64 and arm64 `linuxkit/alpine` images as well as initial support for building arm64 kernels. So we are getting close to full arm64 support.

We want this support to be transparent, ie ideally we can use the same YAML file without any changes to build `x86_64` and `arm64` LinuxKit images (and other archs in the future). The best way to achieve this is using multiarch images/manifests, currently using the [`manifest-tool`](https://github.com/estesp/manifest-tool). The tool pushes a manifest to hub which contains references to to the different images (e.g. for different architectures). I suggest we add the `manifest-tool` the the top-level `Makefile` in `linuxkit`.

## Packages and kernel

Assuming a multiarch `linuxkit/alpine` (see below), for normal packages (and with some minor mods, for the kernel packages) we can use the command line options for `manifest-tool` as the hash is based on the git content hash, so it should be the same on `x86_64` and `arm64`. We need to modify the package build to first push/sign images with the architecture appended, so instead of, e.g.,
```
linuxkit/binfmt:0bde4ebd422099f45c5ee03217413523ad2223e5
```
we would push/sign:
```
linuxkit/binfmt:0bde4ebd422099f45c5ee03217413523ad2223e5-x86_64
linuxkit/binfmt:0bde4ebd422099f45c5ee03217413523ad2223e5-arm64
```
Then in a final step, we'd invoke:
```
DOCKER_CONTENT_TRUST=1 manifest-tool push from-args \
    --ignore-missing \
    --platforms linux/x86_64,linux/arm64 \
    --template linuxkit/binfmt:0bde4ebd422099f45c5ee03217413523ad2223e5-ARCH \
    --target linuxkit/binfmt:0bde4ebd422099f45c5ee03217413523ad2223e5
```
This should create a multiarch image on hub which can be used as before, except it should work both on `x86_64` and `arm64` provided that both images exist. The `--ignore-missing` should allow us to first run `make push` on one platform (which pushes a multiarch image with just one platform) and then, afterwards on the other (which then should repush the image with both architectures. The advantage of this approach is that it requires very little coordination between the architectures and that it can be applied pretty much to all our packages with just changes to `package.mk`.

## Base image

The `linuxkit/alpine` package is slightly different and more awkward. Here, the hash is computed based, primarily, on the hash of the alpine packages mirrored in the image and that means that hash will always be different between the different architectures. To create a multiarch `linuxkit/alpine` package, we need to create a YAML file with the architectures and use the `from-spec` option of `manifest-tool`. This can mostly be scripted as we stash the hashes in the `versions.<arch>` files. However, we need to derive a hash for the manifest "image" and ideally store the resulting hash somewhere.

For the multiarch `linuxkit/alpine`, we can provide a separate make target, given that it already requires some manual steps. We can then use the git tree hash, assuming `version.<arch>` is checked in, and would get a unique hash. The main issue here is that we are back to matching git content hashes to commits and finding out what the `"latest"` base image is. If manifests support labels (and if they were supported by `manifest-tool`) we could tag the manifest image with git commit and repository as we do for other packages...

An alternative would be to always create the multiarch image based on the hash of either the `x86_64` or the `arm64` version. If only "other" image got updated it would still update the manifest so this should work. This means that we have to push the base images with arch suffixes just like proposed for packages.

I think I favour the latter option. It is similar to how packages/kernel is proposed to be handled.

## Assumptions

I'm making a few assumptions:
- ~content trust works with `manifest-tool`~ [it doesn't](https://github.com/linuxkit/linuxkit/issues/2292#issuecomment-318016530)
- we can re-push a new manifest with different content [tested and it works]
- we can use the `ARCH` substitution in tags (all the `manifest-tool` examples use it in the repo name). [WORKS](https://github.com/linuxkit/linuxkit/issues/2292#issuecomment-318045860) 
- the `moby` tool will handle manifest files (including content trust). [WORKS (without content trust)](https://github.com/linuxkit/linuxkit/issues/2292#issuecomment-318038373)
- It's `x86_64` and not `amd64` [it's GOARCH, so `amd64`]
- once we move package builds to CI we can coordinate it sufficiently that on arch is run after the other to avoid a race between pushing the manifest.
- probably a few others...

@justincormack @riyazdf @arm64b @ijc (and others) do you have any comments? would this work? To me it looks doable with relatively little effort and minimum disruption

**Comment by riyazdf** [comment]
we don't have a content trust integration with `manifest-tool` today – though we could (manually with `notary`) sign on the top level manifest list digest (assuming we want the image names to be the same across architectures) and that would work.

I am not too familiar with `manifest-tool` but I imagine integrating the notary library into the upstream code to do this should be fairly straight forward.

**Comment by justincormack** [comment]
I would assume `manifest-tool` might not work directly with eg `DOCKER_CONTENT_TRUST` so you might have to use `notary` to sign. I would check!

Moby tool should be fine at present as it uses `docker`/Docker API. Later we need to fix it - not sure `containerd` has proper manifest support yet.

**Comment by justincormack** [comment]
@riyazdf probably not worth integrating as I think manifest will be in docker cli soon see https://github.com/docker/cli/pull/138

**Comment by ijc** [comment]
For the alpine base I've been wondering about producing a single multiarch mirror image with just hte packages (so no binaries at all) in it as a base image for the arch specific images.

In this scheme the common base would run `apk fetch --arch $arch --recursive -o /mirror/$arch` for all the arches we care about in one build and then use a multistage build to produce an image which contained _only_ `/mirror` (so no `/bin` or anything else) as the end result. Since the various `apk fetch` runes were run back to back they _should_ normally be fairly consistent (in the absence of `apk` functionality to mirror subsets multiple arches simultaneously we could include a check for this if we wanted and rebuild if it fails, although there may be expected drift so maybe that isn't practical). The result would be a `linuxkit/alpine-mirror:SHAxxx` image.

Then the `linuxkit/alpine-ARCH` images would include `FROM linuxkit/alpine-mirror:SHAxxx AS mirror` and `COPY --from mirror /mirror/ARCH /mirror` as well as the usual `apk add -r /mirror -p /out` based runes to produce the output filesystem (based on mirror). That would involve using a few different stages to produce an image based only on the content of the mirror, but not insurmountable, you'd likely need the `FROM mirror` phase and then a `FROM alpine:X.Y` phase to run `apk --out` to produce the input to a `FROM scratch` phase IYSWIM. This would mean that the `linuxkit/alpine-ARCH` packages were reproducibly built from the same mirror image.

The need for `linuxkit/alpine-ARCH` only arises because we want to run things directly in the mirror image, if we were willing to modify all the package builds we could instead arrange for `/mirror/ARCH` to container `/mirror/ARCH/{mirror,bin,etc}` and start each package build with:
```
FROM linuxkit/alpine:SHAxxxx` AS mirror
FROM scratch
COPY --from mirror /mirror/$ARCH /
... continue as usual...
```
Bit ugly that though, although we could combine the two approaches and have the `linuxkit/alpine-ARCH` image do that dance.

On "in the absence of `apk` functionality to mirror multiple arches", perhaps @ncopa can advice if such a thing exists.

**Comment by rn** [comment]
I played with `manifest-tool` and the following yaml:
```
image: linuxkit/alpine:deadbeefdeadbeef
manifests:
  - image: linuxkit/alpine:964e39653b32464f369e52099781b35642a253e5
    platform:
      architecture: arm64
      os: linux
  - image: linuxkit/alpine:34af9cb1990debd17fae6d4198c62ce3910d9908
    platform:
      architecture: amd64
      os: linux
```

```
DOCKER_CONTENT_TRUST=1 manifest-tool push from-spec foo.yml
```
Did not ask me for the notary password and:
```
DOCKER_CONTENT_TRUST=1 docker pull linuxkit/alpine:deadbeefdeadbeef
No trust data for deadbeefdeadbeef
```
So the manifest is not signed. First assumption above is out of the window...

But, overall it seems to work:
```
[arm64 machine] docker run --rm linuxkit/alpine:deadbeefdeadbeef uname -a
Linux 90fd4cfb8800 4.10.0-26-generic #30~16.04.1-Ubuntu SMP Tue Jun 27 09:40:29 UTC 2017 aarch64 Linux

[x86_64 machine] docker run --rm linuxkit/alpine:deadbeefdeadbeef uname -a
Linux 3baffc74a1c1 4.9.36-moby #1 SMP Fri Jul 14 17:47:34 UTC 2017 x86_64 Linux
```

Given that the manifest is just a wrapper around other images, one option might be to modify `moby` to check the signature of the image actually used on a given platform. @justincormack @riyazdf  I have no idea if this is feasible.

**Comment by rn** [comment]
alternatively, we could revert back to using `sha256` until the `docker manifest` support is in...

**Comment by rn** [comment]
I tried (and failed) to derive the `notary` command line from the documentation, but that certainly is the best option until `docker manifest` is available. I think I was able to compile `notary` on arm64 as well so we could make this consistent across platforms

**Comment by rn** [comment]
Created a multiarch image with:
```
image: linuxkit/init:deadbeefdeadbeef
manifests:
  - image: linuxkit/init:838b772355a8690143b37de1cdd4ac5db725271f
    platform:
      architecture: amd64
      os: linux
```
and then  used `linuxkit/init:deadbeefdeadbeef` in `linuxkit.yml` with `-disable-content-trust` and that worked

**Comment by rn** [comment]
Finally, I tagged/pushed the arm64/x86 images from [here](https://github.com/linuxkit/linuxkit/issues/2292#issuecomment-318016530) as `linuxkit/alpine:deadbeefdeadbeef-arm64` and `linuxkit/alpine:deadbeefdeadbeef-amd64` respectively and then used:
```
manifest-tool push from-args     --platforms linux/amd64,linux/arm64 --template linuxkit/alpine:deadbeefdeadbeef-ARCH  --target linuxkit/alpine:deadbeefdeadbeef
```
and it worked with the template substitution.

**Comment by rn** [comment]
The only piece missing now before putting everything together is the manual signing with `notary`

**Comment by rn** [comment]
@ijc In principle,  I like your proposal of creating basically a scratch image with `/mirror` for all architectures. This would have a sha based on content of sorts. Then we could then build the architecture specific ones based off the mirror.

However, the base also includes some go binaries which we `go get` and these are pulled from the network, so we can't use the git tree hash for the architecture specific base images, which kinda leaves us where we are now. (NB, the go binaries are included in the hash calculation).

But maybe I missed something from your proposal...

**Comment by ijc** [comment]
> But maybe I missed something from your proposal...

No, I'd just forgotten about the go bits...

It's possible we could `go get GOARCH=ARCH` a bunch of times in the base image?

Or we could use specific commits of things those things and do them in the per arch images? Normally one uses specific commits to avoid linters externally changing and breaking master, which doesn't quite apply to us since we basically freeze them in stone with the mirror image so that can't happen. We just get the occasional unexpected fail in a new PR when you bump the alpine version instead, but that is controlled and doesn't affect master.

**Comment by ijc** [comment]
`go get -d` in the base (download only) and then `go install` in the arch Dockerfile might work?

**Comment by arm64b** [comment]
@rn I am not very familiar the entire moby software stack. but from my point, seems the only missed is the content trust with manifest-tool, I haven't the time to try that tool, but if we can achieve like:
in arm64 system: `git pull linuxkit/alpine: deadbeaf` will fetch the aarch64 image from the hub, while
`git pull linuxkit/alpine: deadbeef `will fetch the x86_64 image on x86 platform, I think it will be very useful for us to keep all the files intact when supporting multi-archs.  Anyway, this is only IMHO of mine.

**Comment by arm64b** [comment]
> The only piece missing now before putting everything together is the manual signing with notary

Can we integrate the notary signing part into the like Makefile before pushing the image?

**Comment by rn** [comment]
@arm64b yes that is the plan. I got side tracked with some kernel issues elsewhere but @riyazdf is working with @estesp (one of the authors of `manifest-tool`) to get the information out of `manifest-tool` we need to perform a manual `notary add` in the Makefile.

Basically, the `manifest-tool` (and soon the `docker manifest` extension) allows us to have a *single* `linuxkit/alpine` "image" which resolves either to ax86 or to arm64 depending on where it is pulled. This means that then in the kernel compile and in the other packages we don't have to use architecture specific images *and* we can also create multiarch packages, which means that even the YAML files used by `moby` won't be architecture specific, ie we can build a LinuxKit image for x86 and arm64 from the same source YAML. This is what we want for LinuxKit and I think with you work on `linuxkit/alpine` we are very close to that.

The way this works is that you push say `linuxkit/foo:<hash>-amd64` and `linuxkit/foo:<hash>-arm64` to hub and then push a manifest list to hub as `linuxkit/foo:<hash>` and the manifest says, of pulled on `arm64` pull `linuxkit/foo:<hash>-arm64` and if pulled on `amd64`, pull `linuxkit/foo:<hash>-amd64`.

This works today (as per my experiments above), however we also need content trust, ie signed images which are verified on pull. While we can pushed signed images for `linuxkit/foo:<hash>-amd64` and `linuxkit/foo:<hash>-arm64`, we currently can't sign the manifest because for that we need the `sha256` digest and the length. We have the digest, but not yet the length. The plan is to modify `manifest-tool inspect` to give us the length. Then we can use `notary add` to sign it.

**Comment by arm64b** [comment]
> The way this works is that you push say linuxkit/foo:<hash>-amd64 and linuxkit/foo:<hash>-arm64 to hub and then push a manifest list to hub as linuxkit/foo:<hash> and the manifest says, of pulled on arm64 pull linuxkit/foo:<hash>-arm64 and if pulled on amd64, pull linuxkit/foo:<hash>-amd64.

probably I am missing some context here, but my question here is: is the suffix '-amd64' or '-aarch64'  necessary for the push or pull operation? 

Now I am trying the codes in master branch of LinuxKit git repo to see if those codes can run smoothly on ARM64, but encounter some issues. I am narrowing down the issue. For the multi-arch support, let me know if I can be some kind help such as some experiments on ARM64. Hopefully we'll finalized a perfect solution to accomplish that in terms of multi-arch support for LinuxKit, then move to some kinds of advanced topic.

**Comment by rn** [comment]
the suffixes `-amd64` and `-arm64` are not necessary but they allow us to use the `push from-args` mode of `manifest-tool` using the `ARCH` template

**Comment by ncopa** [comment]
> @ijc On "in the absence of apk functionality to mirror multiple arches", perhaps @ncopa can advice if such a thing exists.

There are no support for multiple arches (eg. `apk fetch --arch x86,x86_64,armhf ...`) but it should be trivial to implement something like that as a shell function.

(seems like `apk fetch --arch otherarch` does not work at all)

**Comment by rn** [comment]
I have a modified version of `manifest-tool` at: https://github.com/rn/manifest-tool/tree/notary which prints out the size:
```
./manifest-tool --username rneugeba --password <password> push from-spec ~/tmp/foo.yml
Digest: sha256:ace64c1b634887f28c3633fb34787e2b87e683f9d3a9eae7ec1a1e7f7cd6fbcc 743
```
Then with `notary` we can:
```
notary -s https://notary.docker.io -d ~/.docker/trust addhash -p docker.io/linuxkit/alpine deadbeef 743 --sha256 ace64c1b634887f28c3633fb34787e2b87e683f9d3a9eae7ec1a1e7f7cd6fbcc -r targets/rolf
```
(the `manifest-tool` invocation pushed to `linuxkit/alpine:deadbeef`) and a subsequent `DOCKER_CONTENT_TRUST=1 docker pull linuxkit/alpine:deadbeef` seems to work after reseting notary:
```
notary reset -d ~/.docker/trust docker.io/linuxkit/alpine --all
```

Getting there...

**Comment by ijc** [comment]
@ncopa 

> There are no support for multiple arches (eg. apk fetch --arch x86,x86_64,armhf ...) but it should be trivial to implement something like that as a shell function.

I suppose you mean `for arch in <...> ; do apk fetch --arch $arch...` to do them serially but what I meant was to mirror them all at once so as to get a consistent snapshot across the repo for all (interesting) arches, if you do them serially then there is opportunities for skew (e.g. if there are large package downloads).

**Comment by rn** [comment]
@ijc `apk fetch --arch $arch` does not work. It seems to ignore the arch and pulls identical images. I tried because I wanted to implement your proposal and then realiased after that @ncopa already mentioned above that it doesn't work...

**Comment by ncopa** [comment]
> @ijc I suppose you mean for arch in <...> ; do apk fetch --arch $arch... to do them serially but what I meant was to mirror them all at once so as to get a consistent snapshot across the repo for all (interesting) arches, if you do them serially then there is opportunities for skew (e.g. if there are large package downloads).

Downloading in parallel does not really solve that problem since mirrors may be updated (and dependencies may change) between the time when first package and last package is downloaded.

To do it properly (without doing the snapshot on server side) we'd need to:
1) download index for all arches
2) calculate the dependencies for each arch and generate a list of wanted packages
3) verify that version number is the same on all arches
4) download all packages
5) verify checksum with the index

That way we can verify that we actually get a consistent snapshot and error (or retry) if we don't.

But as @rn said, we need to fix `apk fetch --arch ...` first.

**Comment by ncopa** [comment]
I wonder if it may be worth making a small Lua script (or python) to do it. That way we could also download in parallel (and potentially reduce risk for inconsistency).

**Comment by rn** [comment]
@ncopa @ijc frankly I'm not too worried about having exactly the same versions of all packages for all archs. We have a way to record the versions of packages in the mirror per arch and that would be checked in. It would be nice if they all have the same version across all archs, but we don't want to be dependent on the speed of different builders etc.

**Comment by ijc** [comment]
> I'm not too worried about having exactly the same versions of all packages for all archs

OK

> we don't want to be dependent on the speed of different builders etc.

TBC: the reason I keep banging on about having a single command which can do a consistent snapshot is so that we aren't dependent on precisely things like that. The current method of building N images for N architectures is the one which suffers from this problem. Doing N calls to `apk fetch --arch` (if/when it is fixed) in a single image build would mitigate that somewhat, but still imperfectly depending on the download size.

**Comment by rn** [comment]
closed via #2350
